aloe/registry.py

```diff
@@ -449,3 +454,111 @@ def decorator(self, function, **kwargs):
+        """Register a placeholder to be used on the extended_step."""
+
+        if len(args) > 1:
+            self.multi_expression_placeholders[placeholder] = args
+        else:
+            self.single_expression_placeholders[placeholder] = args[0]
+
+    def _replace_placeholders(self, sentence):
+        """
+        Replace placeholders with their associated expressions.
+        Returns a list with at least one sentences to the product of all
+        the expressions when one or more multiple-expressions are used.
+        `replace` is use instead of `format` as the latter interferes with
+        defining complex custom regexes that contains {}.
+        """
+
+        for placeholder, expression in \
+                self.single_expression_placeholders.iteritems():
+            sentence = sentence.replace(self.CLOSURE % placeholder, expression)
+
+        sentences = [sentence]
+
+        for placeholder in self.multi_expression_placeholders.keys():
+            sentences = self._replace_multi_expression_placeholder(
+                sentences, placeholder
+            )
+
+        return sentences
+
+    # pylint:disable=invalid-name
+    def _replace_multi_expression_placeholder(self, sentences, placeholder):
+        """
+        Replace a placeholder that is associated with multiple expressions.
+        """
+
+        placeholder_str = self.CLOSURE % placeholder
+
+        if placeholder_str not in sentences[0]:
+            return sentences
+
+        expressions = self.multi_expression_placeholders[placeholder]
+        new_sentences = []
+
+        for expression in expressions:
+            for sentence in sentences:
+                new_sentences.append(
+                    sentence.replace(placeholder_str, expression)
+                )
+
+        return new_sentences
+    # pylint:enable=invalid-name
+
+    def extended_step(self, sentence):
+        """
+        Creates one or multiple step definitions and associate them to the same
+        function.
+        The sentence can contain placeholders that are replaced by common
+        expressions/regexes.
+        A placeholder can be associated with multiple regexes creating in that
+        way multiple step definitions.
+        Placeholders are case sensitive and are enclosed in {}.
+        Common placeholders:
+            {STRING}
+            {NUMBER}
+            {NON_CAPTURING_STRING}
+            {NON_CAPTURING_NUMBER}
+        """
+
+        def decorator(func):
+            """Register a function as a step using the parsed sentence."""
+
+            sentences = self._replace_placeholders(sentence)
+
+            for parsed_sentence in sentences:
+                step(parsed_sentence)(func)
+
+        return decorator
+
+
+EXTENDED_STEP = ExtendedStep()
+
+# These are functions, not constants
+# pylint:disable=invalid-name
+register_placeholder = EXTENDED_STEP.register_placeholder
+extended_step = EXTENDED_STEP.extended_step
```

Comment:
This makes API stink. (What if you add another way of parsing steps, `extended_extended_step`?)

A better way would be to do `@step("I have a NUMBER", parser=step.NICE (or step.REGEX))`, or again look at Behave.